src/MediatR/Mediator.cs

```diff
@@ -226,6 +256,25 @@ private IEnumerable<TWrapper> GetNotificationHandlers<TWrapper>(object notificat
                 .ToList();
         }
 
+        private Task<TResponse> GetPipeline<TResponse>(object request, RequestHandlerDelegate<TResponse> invokeHandler)
+        {
+            var requestType = request.GetType();
+            var method = CreatePipelineMethod.MakeGenericMethod(requestType, typeof(TResponse));

```

Comment:
My first thought was to pass a state machine into the pipeline which would be executed on every step. This would work and flip all the closed-over instances into the state machine which is passed as an input argument but this call chain pollutes the stack trace quite a bit.

If you really want to do this right you have to do some code gen with expression trees.
Had a call with Daniel (Marbach) this monday where we concluded this. As he has created something very similar for NServiceBus.